File: Challenger/HomeViewController.m
//
//  HomeViewController.m
//  Challenger
//
//  Created by Sreenivas Appasani on 11/5/19.
//  Copyright © 2019 Sreenivas Appasani. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "HomeViewController.h"
#import "CGECollectionViewCell.h"

@interface HomeViewController () <UICollectionViewDataSource>
@property (weak, nonatomic) IBOutlet UICollectionView *ChallengeCV;
@property (strong, nonatomic) NSMutableArray *challenges;
@end

@implementation HomeViewController

-(void) viewDidLoad {
    self.challenges = [[NSMutableArray alloc] initWithArray:@[@"foo", @"bar", @"boo"]];
    self.ChallengeCV.dataSource = self;
    
}

- (nonnull __kindof UICollectionViewCell *)collectionView:(nonnull UICollectionView *)collectionView cellForItemAtIndexPath:(nonnull NSIndexPath *)indexPath {
    static BOOL nibMyCellloaded = NO;
    static NSString *cellIdentifier = @"CGECell";
    
    if(!nibMyCellloaded)
    {
        UINib *nib = [UINib nibWithNibName:@"CGECollectionViewCell" bundle: nil];
        [collectionView registerNib:nib forCellWithReuseIdentifier:cellIdentifier];
        nibMyCellloaded = YES;
    }
    
    CGECollectionViewCell *cell = (CGECollectionViewCell *)[collectionView dequeueReusableCellWithReuseIdentifier:cellIdentifier forIndexPath:indexPath];
    [cell layoutIfNeeded];
    return cell;
}

- (NSInteger)collectionView:(nonnull UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return self.challenges.count;
}

@end

File: Challenger/HomeViewController.swift
//
//  HomeViewController.swift
//  Challenger
//
//  Created by Sreeni Appasani on 12/1/19.
//  Copyright © 2019 Sreenivas Appasani. All rights reserved.
//

import Foundation
import UIKit
import Alamofire

class HomeViewController : UIViewController, UICollectionViewDataSource {
    @IBOutlet weak var challengeCV: UICollectionView!
    
    private var challenges: [Any] = []

    override func viewDidLoad() {
        Challenge.getChallenges { challengesArr in
            self.challenges = challengesArr
//            self.tableView.reloadData()
            DispatchQueue.main.async {
                self.challengeCV.reloadData()
            }
        }
//        challenges = ["foo", "bar", "boo"]
        challengeCV.dataSource = self
    }

    static var collectionViewNibMyCellloaded = false
    static let collectionViewCellIdentifier = "CGECell"

    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {

        if !HomeViewController.collectionViewNibMyCellloaded {
            let nib = UINib(nibName: "CGECollectionViewCell", bundle: nil)
            collectionView.register(nib, forCellWithReuseIdentifier: HomeViewController.collectionViewCellIdentifier)
            HomeViewController.collectionViewNibMyCellloaded = true
        }

        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: HomeViewController.collectionViewCellIdentifier, for: indexPath) as? CGECollectionViewCell
        cell?.layoutIfNeeded()
        return cell!
    }

    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return challenges.count
    }
}
